### ISSUE TITLE
Support SSO/SCIM configuration

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/terraform-provider-materialize#430]
Via Frontegg, we support configuring SSO and SCIM. At present, the only supported way to configure SSO and SCIM is via clickops in the account settings panel, but many users are likely to want to configure SSO and SCIM via Terraform provider.

It should be straightforward to expose resources for the following in the Terraform provider via the end user Frontegg API (admin.cloud.materialize.com):

  * [x] SSO configuration
      * [API reference](https://docs.frontegg.com/reference/ssoconfigurationcontrollerv1_createssoconfiguration)
  * [ ] Groups
    * [ ] [API reference](https://docs.frontegg.com/reference/groupscontrollerv1_creategroup)
  * [x] SCIM
    * No documentation, but sniffing around the account settings portal in the network inspector indicates an API around https://admin.cloud.materialize.com/frontegg/directory/resources/v1/configurations/scim2

cc @bobbyiliev

**Comment by bobbyiliev** [comment]
Just started looking into the SSO configuration, and currently debating between two approaches:

### Case 1: one large resource that combines all SSO API endpoints

```hcl
resource "materialize_sso_config" "example_sso" {
  enabled             = true
  sso_endpoint        = "https://sso.example.com"
  public_certificate  = "cert_data"
  sign_request        = true
  type                = "saml"
  oidc_client_id      = "client-id"
  oidc_secret         = "client-secret"

  domains = ["example.com", "example.org"] # Separate Domains API endpoint
  groups = [                               # Separate Groups API endpoint
    {
      name = "group1"
      role_ids = ["role1", "role2"]        # Separate Roles API endpoint
    },
    {
      name = "group2"
      role_ids = ["role2"]
    },
  ]
}
```

With this, the `domains` and `groups` will be part of the `materialize_sso_config` resource. This approach keeps everything related to the SSO configuration in one place.

### Case 2: Separate Resource for Domains and Groups

In this case, we will have separate resources for the SSO configuration, domains, and groups. The Terraform configuration might be split like this:

```hcl
resource "materialize_sso_config" "example_sso" {
  enabled                = true
  sso_endpoint       = "https://sso.example.com"
  public_certificate = "cert_data"
  sign_request        = true
  type                       = "saml"
  ...
}

resource "materialize_sso_domains" "example_domains" {
  sso_config_id = materialize_sso_config.example_sso.id
  domains       = ["example.com", "example.org"]
}

resource "materialize_sso_groups" "example_group1" {
  sso_config_id = materialize_sso_config.example_sso.id
  name          = "group1"
  role_ids      = ["role1", "role2"]
}

resource "materialize_sso_groups" "example_group2" {
  sso_config_id = materialize_sso_config.example_sso.id
  name          = "group2"
  role_ids      = ["role3"]
}
```

In this approach, the `materialize_sso_domains` and `materialize_sso_groups` resources are separate from the `materialize_sso_config` resource. Each group and set of domains is managed independently, but they reference the SSO configuration via the `sso_config_id` attribute.

---

Open to suggestions, I am personally leaning to case 1 even though it will combine multiple API endpoints and objects.

**Comment by benesch** [comment]
I'd advocate for a hybrid approach. It feels to me like domains should be part of the main SSO configuration (as you can't have an SSO configuration without at least one verified domain), while groups should be independent resources.

Also, wait, do groups really reference an SSO configuration by ID? I don't see that in the Frontegg API documentation... but also the Frontegg API documentation is often incomplete.

**Comment by bobbyiliev** [comment]
> Also, wait, do groups really reference an SSO configuration by ID? I don't see that in the Frontegg API documentation... but also the Frontegg API documentation is often incomplete.

Yes, the link to the groups docs that you've shared in the description is actually for a different type of groups endpoint and not the one associated with the SSO, for the SSO groups the endpoint is `/team/resources/sso/v1/configurations/{configurationId}/groups`:

https://docs.frontegg.com/reference/ssogroupscontrollerv1_createssogroup

As far as I can see, that above endpoint is what's used to do the SSO group <> roles mapping.

**Comment by bobbyiliev** [comment]
Quick update on this one, I've been working on this today and seems like combining the SSO domain + the SSO configuration into 1 resource there are a few annoying issues that are pretty hard to manage in Terraform. Basically handling errors and managing state becomes pretty complex when multiple configurations are bundled into a single resource.

So I've started leaning towards option 2, as separating them would simplify this process. For example, if creating or updating a domain fails, it doesn’t directly impact the SSO configuration resource, making it easier to handle errors and manage the state. I think that it also matches the recommended best practices by Terraform for using separate resources for the different API resources.

**Comment by benesch** [comment]
> > Also, wait, do groups really reference an SSO configuration by ID? I don't see that in the Frontegg API documentation... but also the Frontegg API documentation is often incomplete.
> 
> Yes, the link to the groups docs that you've shared in the description is actually for a different type of groups endpoint and not the one associated with the SSO, for the SSO groups the endpoint is `/team/resources/sso/v1/configurations/{configurationId}/groups`:
> 
> https://docs.frontegg.com/reference/ssogroupscontrollerv1_createssogroup
> 
> As far as I can see, that above endpoint is what's used to do the SSO group <> roles mapping.

Ah! Good spot. 

> So I've started leaning towards option 2, as separating them would simplify this process. For example, if creating or updating a domain fails, it doesn’t directly impact the SSO configuration resource, making it easier to handle errors and manage the state. I think that it also matches the recommended best practices by Terraform for using separate resources for the different API resources.

SGTM.

**Comment by benesch** [comment]
Ah! I realized my confusion earlier around groups. What you've implemented in #442 is about SSO group <-> roles mapping for SSO JIT provisioning (see step 7 [here](https://docs.frontegg.com/docs/saml-idp-instructions)), while there's a totally separate group feature for _SCIM_ provisioning (see [these docs](https://docs.frontegg.com/docs/groups), whose API I linked you at above.

What you've implemented in #442 looks good and self contained, but we'll _also_ want to add resources/data sources for SCIM and SCIM groups, as that's what we'll need to support the customer.
